### ISSUE TITLE
VB compiler incorrectly treats a parenthesized receiver as a variable

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#4687]
1. Run the following code

``` vb
Imports System.Threading.Tasks
Module Module1
    Sub Main()
        System.Console.WriteLine("---")
        Dim s1 = {New S1 With {.F1 = 1}}
        System.Console.WriteLine(s1(0).F1)
        Test1(s1)
        System.Console.WriteLine(s1(0).F1)
        System.Threading.Thread.Sleep(4000)
    End Sub
    Sub Test1(x() As S1)
        Call (x(0)).MutatingCall()
    End Sub
End Module
Interface I1
    Sub MutatingCall()
End Interface
Structure S1
    Implements I1
    Public F1 As Integer
    Public Sub MutatingCall() Implements I1.MutatingCall
        System.Console.WriteLine("S1.MutatingCall")
        F1 += 1
    End Sub
End Structure
```
## Expected Behavior: Prints

1
S1.MutatingCall
1
## Actual Behavior: Prints

1
S1.MutatingCall
2

---

This was migrated from TFS/DevDiv 1040093

**Comment by gafter** [comment]
@gafter wrote

> This issue should be recorded in the compiler specification before it is closed.

@AnthonyDGreen wrote

> This should be fixed.
>
It's an inviolable law of Visual Basic that parenthesizing an expression forces it to be classified as a value (reclassifying it if necessary). The is the whole reason why parenthesizing a ByRef argument prevents mutation or copy-back to the argument.
>
The native compiler does the correct thing, the current "Roslyn" behavior is wrong.

@AlekseyTs wrote

> I agree, this is a breaking change in Roslyn, which should be fixed. Note, that we need to make sure that the fix also covers Elvis operator.

@AlekseyTs wrote

> Resolved with changeset 1357846: VB Async stack spilling: properly handle spilling of receivers.

@gafter wrote

> Aleksey: Using the latest sources, this still repros as described in repro steps.

@AlekseyTs wrote

> This is expected, see comments in related bug #1026678.

@gafter wrote

> Until we have either an updated language specification or a note in the compiler specification describing the variance from native and spec this remains an open issue.
>
I recommend adding that email thread as a note to the compiler specification.

@AlekseyTs wrote

> Lucian, could you please update the language spec?

**Comment by gafter** [comment]
This issue is now tracked at https://github.com/dotnet/vblang/issues/51
